presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java

```diff
@@ -477,16 +489,15 @@ public CounterStat getHighMemorySplitSource()
 
                 results.add(new HivePartitionMetadata(hivePartition, Optional.of(partition), partitionSchemaDifference.build(), encryptionInformation));
             }
-            Map<String, String> map = partitionsNotReadable.build();
-            Multimap<String, String> multiMap = HashMultimap.create();
-            for (java.util.Map.Entry<String, String> entry : map.entrySet()) {
-                multiMap.put(entry.getValue(), entry.getKey());
-            }
-            for (java.util.Map.Entry<String, Collection<String>> entry : multiMap.asMap().entrySet()) {
-                warningCollector.add(new PrestoWarning(
-                        PARTITION_NOT_READABLE,
-                        format("Table '%s' has '%s' out of '%s' partitions: '%s' not readable: %s", tableName, entry.getValue().size(), partitionBatch.size(), entry.getValue(), entry.getKey())));
+            String prestoWarningMessage = format("Table '%s' has '%s' out of '%s' partitions unreadable: ", tableName, partitionsNotReadable.values().stream().mapToInt(Set::size).sum(), partitionBatch.size());
+            for (Entry<String, Set<String>> entry : partitionsNotReadable.entrySet()) {
+                if (entry.getValue().size() > 2)
+                    prestoWarningMessage += "\n" + entry.getKey().toString() + ": " + entry.getValue().iterator().next() + ", " + entry.getValue().iterator().next() + " ...";
+                else prestoWarningMessage += "\n" + entry.getKey().toString() + ": " + entry.getValue().toString();
             }
+            if (reasonsSkipped > 3)
+                prestoWarningMessage += "\n...";
```

Comment:
Looks a bit hacky, but I tested it for all possible numbers of errors and partitions, and it works.